ws-server/src/server.rs

```diff
@@ -307,102 +324,135 @@ async fn background_task(
 				let d = std::mem::take(&mut data);
 				let resources = &resources;
 				let methods = &methods;
-				let tx2 = tx.clone();
+				let sink = sink.clone();
 
 				let fut = async move {
 					// Batch responses must be sent back as a single message so we read the results from each
 					// request in the batch and read the results off of a new channel, `rx_batch`, and then send the
 					// complete batch response back to the client over `tx`.
 					let (tx_batch, mut rx_batch) = mpsc::unbounded();
+					let sink_batch = MethodSink::new_with_limit(tx_batch, max_request_body_size);
 					if let Ok(batch) = serde_json::from_slice::<Vec<Request>>(&d) {
 						tracing::debug!("recv batch len={}", batch.len());
 						tracing::trace!("recv: batch={:?}", batch);
 						if !batch.is_empty() {
-							let methods_stream = stream::iter(batch.into_iter().filter_map(|req| {
-								methods.execute_with_resources(
-									&tx_batch,
-									req,
-									conn_id,
-									resources,
-									max_request_body_size,
-								)
-							}));
-
-							let results = methods_stream
-								.for_each_concurrent(None, |item| item)
-								.then(|_| {
-									rx_batch.close();
-									collect_batch_response(rx_batch)
-								})
-								.await;
-
-							if let Err(err) = tx2.unbounded_send(results) {
+							join_all(batch.into_iter().filter_map(move |req| {
+								match methods.execute_with_resources(&sink_batch, req, conn_id, resources) {
+									Ok((name, MethodResult::Sync(success))) => {
+										middleware.on_result(name, success, request_start);
+										None
+									}
+									Ok((name, MethodResult::Async(fut))) => Some(async move {
+										let success = fut.await;
+										middleware.on_result(name, success, request_start);
+									}),
+									Err(name) => {
+										middleware.on_result(name.as_ref(), false, request_start);
+										None
+									}
+								}
+							}))
+							.await;
+
+							rx_batch.close();
+							let results = collect_batch_response(rx_batch).await;
+
+							if let Err(err) = sink.send_raw(results) {
 								tracing::error!("Error sending batch response to the client: {:?}", err)
+							} else {
+								middleware.on_response(request_start);
 							}
 						} else {
-							send_error(Id::Null, &tx2, ErrorCode::InvalidRequest.into());
+							sink.send_error(Id::Null, ErrorCode::InvalidRequest.into());
+							middleware.on_response(request_start);
 						}
 					} else {
 						let (id, code) = prepare_error(&d);
-						send_error(id, &tx2, code.into());
+						sink.send_error(id, code.into());
+						middleware.on_response(request_start);
 					}
 				};
 
 				method_executors.add(Box::pin(fut));
 			}
-			_ => send_error(Id::Null, &tx, ErrorCode::ParseError.into()),
+			_ => {
+				sink.send_error(Id::Null, ErrorCode::ParseError.into());
+			}
 		}
-	}
+	};
+
+	middleware.on_disconnect();
 
 	// Drive all running methods to completion
 	method_executors.await;
```

Comment:
Just a note that we should perhaps add a comment or something to make it clear to our future selves that we souldn't return early before this await runs?